File: src/lib/stores/personality.ts
import { writable, get } from 'svelte/store';
import { invoke } from '@tauri-apps/api/core';

export interface RoleInfo {
	code: string;
	name: string;
	available: boolean;
}

export interface PersonalityInfo {
	agents_path: string | null;
	personality_loaded: boolean;
	alc_loaded: boolean;
	active_role: string | null;
	roles: RoleInfo[];
}

export const personalityInfo = writable<PersonalityInfo | null>(null);
export const activeRole = writable<string | null>(null);
export const systemPrompt = writable<string | null>(null);
export const personalityEnabled = writable(true);

export async function loadPersonalityInfo(): Promise<void> {
	try {
		const info = await invoke<PersonalityInfo>('get_personality_info');
		personalityInfo.set(info);

		// Default to 'dna' role if available
		if (!get(activeRole) && info.roles.some(r => r.code === 'dna' && r.available)) {
			activeRole.set('dna');
		}

		// Build system prompt
		await rebuildSystemPrompt();
	} catch (e) {
		console.warn('[personality] failed to load info:', e);
	}
}

export async function setRole(roleCode: string | null): Promise<void> {
	activeRole.set(roleCode);
	await rebuildSystemPrompt();
}

export async function rebuildSystemPrompt(): Promise<void> {
	if (!get(personalityEnabled)) {
		systemPrompt.set(null);
		return;
	}

	try {
		const role = get(activeRole);
		const prompt = await invoke<string>('build_system_prompt', { role });
		systemPrompt.set(prompt);
	} catch (e) {
		console.warn('[personality] failed to build system prompt:', e);
		systemPrompt.set(null);
	}
}

export function togglePersonality(): void {
	const current = get(personalityEnabled);
	personalityEnabled.set(!current);
	if (!current) {
		rebuildSystemPrompt();
	} else {
		systemPrompt.set(null);
	}
}

File: src-tauri/src/personality.rs
use serde::Serialize;
use std::path::PathBuf;

const AGENTS_PATHS: &[&str] = &[
    // Embedded in a torch project
    ".hecate/agents",
    // Development path (relative to home)
    "work/github.com/hecate-social/hecate-agents",
];

const PERSONALITY_FILE: &str = "PERSONALITY.md";
const ALC_FILE: &str = "philosophy/HECATE_ALC.md";

const ROLE_FILES: &[(&str, &str, &str)] = &[
    ("dna", "Discovery & Analysis", "philosophy/HECATE_DISCOVERY_N_ANALYSIS.md"),
    ("anp", "Architecture & Planning", "philosophy/HECATE_ARCHITECTURE_N_PLANNING.md"),
    ("tni", "Testing & Implementation", "philosophy/HECATE_TESTING_N_IMPLEMENTATION.md"),
    ("dno", "Deployment & Operations", "philosophy/HECATE_DEPLOYMENT_N_OPERATIONS.md"),
];

#[derive(Debug, Serialize, Clone)]
pub struct RoleInfo {
    pub code: String,
    pub name: String,
    pub available: bool,
}

#[derive(Debug, Serialize, Clone)]
pub struct PersonalityInfo {
    pub agents_path: Option<String>,
    pub personality_loaded: bool,
    pub alc_loaded: bool,
    pub active_role: Option<String>,
    pub roles: Vec<RoleInfo>,
}

fn find_agents_dir() -> Option<PathBuf> {
    let home = dirs::home_dir()?;

    for rel_path in AGENTS_PATHS {
        let candidate = home.join(rel_path);
        if candidate.join(PERSONALITY_FILE).exists() {
            return Some(candidate);
        }
    }

    // Also check current working directory
    if let Ok(cwd) = std::env::current_dir() {
        let candidate = cwd.join(".hecate/agents");
        if candidate.join(PERSONALITY_FILE).exists() {
            return Some(candidate);
        }
    }

    None
}

fn read_file(path: &PathBuf) -> Option<String> {
    std::fs::read_to_string(path).ok().filter(|s| !s.trim().is_empty())
}

#[tauri::command]
pub fn get_personality_info() -> PersonalityInfo {
    let agents_dir = find_agents_dir();

    let mut info = PersonalityInfo {
        agents_path: agents_dir.as_ref().map(|p| p.to_string_lossy().to_string()),
        personality_loaded: false,
        alc_loaded: false,
        active_role: None,
        roles: Vec::new(),
    };

    if let Some(dir) = &agents_dir {
        info.personality_loaded = dir.join(PERSONALITY_FILE).exists();
        info.alc_loaded = dir.join(ALC_FILE).exists();

        for (code, name, file) in ROLE_FILES {
            info.roles.push(RoleInfo {
                code: code.to_string(),
                name: name.to_string(),
                available: dir.join(file).exists(),
            });
        }
    }

    info
}

#[tauri::command]
pub fn build_system_prompt(role: Option<String>) -> Result<String, String> {
    let agents_dir = find_agents_dir()
        .ok_or_else(|| "hecate-agents not found".to_string())?;

    let mut parts: Vec<String> = Vec::new();

    // 1. Personality
    if let Some(personality) = read_file(&agents_dir.join(PERSONALITY_FILE)) {
        parts.push(personality);
    }

    // 2. ALC overview
    if let Some(alc) = read_file(&agents_dir.join(ALC_FILE)) {
        parts.push(alc);
    }

    // 3. Active role
    if let Some(role_code) = role {
        for (code, _, file) in ROLE_FILES {
            if *code == role_code {
                if let Some(role_content) = read_file(&agents_dir.join(file)) {
                    parts.push(role_content);
                }
                break;
            }
        }
    }

    if parts.is_empty() {
        return Err("No personality files found".to_string());
    }

    Ok(parts.join("\n\n---\n\n"))
}

#[tauri::command]
pub fn list_roles() -> Vec<RoleInfo> {
    let agents_dir = find_agents_dir();

    ROLE_FILES.iter().map(|(code, name, file)| {
        RoleInfo {
            code: code.to_string(),
            name: name.to_string(),
            available: agents_dir.as_ref()
                .map(|d| d.join(file).exists())
                .unwrap_or(false),
        }
    }).collect()
}
